modules/audio_device/win/core_audio_output_win.cc

```diff
@@ -337,9 +341,29 @@ bool CoreAudioOutput::OnDataCallback(uint64_t device_frequency) {
   // `audio_data`. The playout latency is not updated for each callback.
   fine_audio_buffer_->GetPlayoutData(
       rtc::MakeArrayView(reinterpret_cast<int16_t*>(audio_data),
-                         num_requested_frames * format_.Format.nChannels),
+                         num_requested_frames *
+                             // RingRTC change to ensure that only 1 or 2
+                             // channels are read from the buffer.
+                         std::min((UINT16)2, format_.Format.nChannels)),
       latency_ms_);
 
+  // RingRTC change to expand 2 channels to multiple channels.
+  if (format_.Format.nChannels > 2) {
+    UINT32 bytes_per_frame = format_.Format.nChannels * sizeof(int16_t);
+    UINT32 bytes_per_stereo_frame = 2 * sizeof(int16_t);
+
+    // Shift each stereo frame to start on the render frame's stride and
+    // set the remaining channels to zero (silence).
+    int16_t* audio_data_ptr = (int16_t*)audio_data;
+    for (UINT32 i = num_requested_frames - 1; i > 0; --i) {
+      memcpy(&audio_data_ptr[i * format_.Format.nChannels],
+             &audio_data_ptr[i * 2], bytes_per_stereo_frame);
```

Comment:
Nitpick: potentially-overlapping copies should be done with `memmove`, not `memcpy` (this is overlapping for i = 1|n = 3).

…but moreover you're doing a backwards loop rather than a forwards loop and that destroys data; this kind of loop has to be done forwards when shrinking in-place.

Reply:
This isn't shrinking though, it is expanding ? I'll look at it again though.

Reply:
Yeah, I read the loop backwards, but "i = 1|n=3" is still overlapping.